File: core/src/test/java/icyllis/modernui/test/TestRectanglePacker.java
/*
 * Modern UI.
 * Copyright (C) 2019-2023 BloCamLimb. All rights reserved.
 *
 * Modern UI is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Modern UI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Modern UI. If not, see <https://www.gnu.org/licenses/>.
 */

package icyllis.modernui.test;

import icyllis.modernui.graphics.Rect;
import icyllis.modernui.graphics.RectanglePacker;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.core.config.Configurator;
import org.lwjgl.glfw.GLFW;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Random;

public class TestRectanglePacker {

    public static final int WIDTH =  1024;
    public static final int HEIGHT = 1024;

    public static void main(String[] args) throws IOException {
        Configurator.setRootLevel(Level.INFO);

        var bm1 = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        var g2d1 = bm1.createGraphics();

        var bm2 = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        var g2d2 = bm2.createGraphics();

        var bm3 = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        var g2d3 = bm3.createGraphics();

        var bm4 = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        var g2d4 = bm4.createGraphics();

        var bm5 = new BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB);
        var g2d5 = bm5.createGraphics();

        var packer1 = RectanglePacker.make(WIDTH, HEIGHT, RectanglePacker.ALGORITHM_SKYLINE);
        var packer2 = RectanglePacker.make(WIDTH, HEIGHT, RectanglePacker.ALGORITHM_HORIZON);
        var packer3 = RectanglePacker.make(WIDTH, HEIGHT, RectanglePacker.ALGORITHM_HORIZON_OLD);
        var packer4 = RectanglePacker.make(WIDTH, HEIGHT, RectanglePacker.ALGORITHM_BINARY_TREE);
        var packer5 = RectanglePacker.make(WIDTH, HEIGHT, RectanglePacker.ALGORITHM_POWER2_LINE);
        var random = new Random();

        double time1 = 0;
        double time2 = 0;
        double time3 = 0;
        double time4 = 0;
        double time5 = 0;

        GLFW.glfwInit();

        int n1 = 0, n2 = 0, n3 = 0, n4 = 0, n5 = 0, fails = 0;
        while (fails < 3000) {
            int w = random.nextInt(8, 33);
            int h = random.nextInt(12, 37);
            var col = new Color(random.nextInt(0x1000000));
            {
                var rect = new Rect(0, 0, w, h);
                double start = GLFW.glfwGetTime();
                if (packer1.addRect(rect)) {
                    g2d1.setColor(col);
                    g2d1.drawRect(rect.left, rect.top, rect.width() - 1, rect.height() - 1);
                    ++n1;
                } else {
                    ++fails;
                }
                time1 += GLFW.glfwGetTime() - start;
            }
            {
                var rect = new Rect(0, 0, w, h);
                double start = GLFW.glfwGetTime();
                if (packer2.addRect(rect)) {
                    g2d2.setColor(col);
                    g2d2.drawRect(rect.left, rect.top, rect.width() - 1, rect.height() - 1);
                    ++n2;
                } else {
                    ++fails;
                }
                time2 += GLFW.glfwGetTime() - start;
            }
            {
                var rect = new Rect(0, 0, w, h);
                double start = GLFW.glfwGetTime();
                if (packer3.addRect(rect)) {
                    g2d3.setColor(col);
                    g2d3.drawRect(rect.left, rect.top, rect.width() - 1, rect.height() - 1);
                    ++n3;
                } else {
                    ++fails;
                }
                time3 += GLFW.glfwGetTime() - start;
            }
            {
                var rect = new Rect(0, 0, w, h);
                double start = GLFW.glfwGetTime();
                if (packer4.addRect(rect)) {
                    g2d4.setColor(col);
                    g2d4.drawRect(rect.left, rect.top, rect.width() - 1, rect.height() - 1);
                    ++n4;
                } else {
                    ++fails;
                }
                time4 += GLFW.glfwGetTime() - start;
            }
            {
                var rect = new Rect(0, 0, w, h);
                double start = GLFW.glfwGetTime();
                if (packer5.addRect(rect)) {
                    g2d5.setColor(col);
                    g2d5.drawRect(rect.left, rect.top, rect.width() - 1, rect.height() - 1);
                    ++n5;
                } else {
                    ++fails;
                }
                time5 += GLFW.glfwGetTime() - start;
            }
        }
        System.out.println("Algorithm Skyline:");
        System.out.println(n1 + " rectangles");
        System.out.println("Coverage: " + packer1.getCoverage() + " (↑)");
        System.out.printf("Took %d microseconds (↓)\n", (int) (time1 * 1000000));

        System.out.println("-".repeat(20));

        System.out.println("Algorithm Horizontal Line:");
        System.out.println(n2 + " rectangles");
        System.out.println("Coverage: " + packer2.getCoverage() + " (↑)");
        System.out.printf("Took %d microseconds (↓)\n", (int) (time2 * 1000000));

        System.out.println("-".repeat(20));

        System.out.println("Algorithm Horizontal Line Memoryless:");
        System.out.println(n3 + " rectangles");
        System.out.println("Coverage: " + packer3.getCoverage() + " (↑)");
        System.out.printf("Took %d microseconds (↓)\n", (int) (time3 * 1000000));

        System.out.println("-".repeat(20));

        System.out.println("Algorithm Binary Tree:");
        System.out.println(n4 + " rectangles");
        System.out.println("Coverage: " + packer4.getCoverage() + " (↑)");
        System.out.printf("Took %d microseconds (↓)\n", (int) (time4 * 1000000));

        System.out.println("-".repeat(20));

        System.out.println("Algorithm Power 2 Line:");
        System.out.println(n5 + " rectangles");
        System.out.println("Coverage: " + packer5.getCoverage() + " (↑)");
        System.out.printf("Took %d microseconds (↓)\n", (int) (time5 * 1000000));

        ImageIO.write(bm1, "png", new File("rect_packer1.png"));
        ImageIO.write(bm2, "png", new File("rect_packer2.png"));
        ImageIO.write(bm3, "png", new File("rect_packer3.png"));
        ImageIO.write(bm4, "png", new File("rect_packer4.png"));
        ImageIO.write(bm5, "png", new File("rect_packer5.png"));

        GLFW.glfwTerminate();
    }
}

File: core/src/test/java/icyllis/modernui/test/TestRectanglePacker.kt
/*
 * Modern UI.
 * Copyright (C) 2019-2023 BloCamLimb. All rights reserved.
 *
 * Modern UI is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Modern UI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Modern UI. If not, see <https://www.gnu.org/licenses/>.
 */

package icyllis.modernui.test

import icyllis.arc3d.core.Rect2i
import icyllis.arc3d.core.RectanglePacker
import icyllis.modernui.graphics.MathUtil
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.core.config.Configurator
import org.lwjgl.glfw.GLFW
import java.awt.Color
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import java.io.File
import java.util.*
import javax.imageio.ImageIO

const val WIDTH = 1024
const val HEIGHT = 1024

fun main() {
    System.setProperty("java.awt.headless", "true")
    Configurator.setRootLevel(Level.INFO)

    val packer1 = TestRectanglePacker(RectanglePacker.ALGORITHM_SKYLINE)
    val packer2 = TestRectanglePacker(RectanglePacker.ALGORITHM_HORIZON)
    val packer3 = TestRectanglePacker(RectanglePacker.ALGORITHM_HORIZON_OLD)
    val packer4 = TestRectanglePacker(RectanglePacker.ALGORITHM_BINARY_TREE)
    val packer5 = TestRectanglePacker(RectanglePacker.ALGORITHM_POWER2_LINE)

    val random = Random()

    GLFW.glfwInit()

    var hue = 0f
    do {
        val w = MathUtil.clamp(20 + (6 * random.nextGaussian()).toInt(), 8, 32)
        val h = MathUtil.clamp(24 + (6 * random.nextGaussian()).toInt(), 12, 36)
        val col = Color(icyllis.arc3d.core.Color.HSVToColor(hue, 1f, 1f))
        hue += 1f / 6f
    } while (
        packer1.add(w, h, col) or
        packer2.add(w, h, col) or
        packer3.add(w, h, col) or
        packer4.add(w, h, col) or
        packer5.add(w, h, col)
    )
    println("Algorithm Skyline:")
    packer1.print()

    println("-".repeat(20))

    println("Algorithm Horizon:")
    packer2.print()

    println("-".repeat(20))

    println("Algorithm Horizon (Old):")
    packer3.print()

    println("-".repeat(20))

    println("Algorithm Binary Tree:")
    packer4.print()

    println("-".repeat(20))

    println("Algorithm Power Of Two:")
    packer5.print()

    ImageIO.write(packer1.bm, "png", File("rect_packer1.png"))
    ImageIO.write(packer2.bm, "png", File("rect_packer2.png"))
    ImageIO.write(packer3.bm, "png", File("rect_packer3.png"))
    ImageIO.write(packer4.bm, "png", File("rect_packer4.png"))
    ImageIO.write(packer5.bm, "png", File("rect_packer5.png"))

    GLFW.glfwTerminate()
}

class TestRectanglePacker(algorithm: Int) {
    private val packer: RectanglePacker
    val bm: BufferedImage
    private val g2d: Graphics2D

    private var time = 0.0
    private var n = 0
    private var fails = 0

    init {
        packer = RectanglePacker.make(WIDTH, HEIGHT, algorithm)
        bm = BufferedImage(WIDTH, HEIGHT, BufferedImage.TYPE_INT_RGB)
        g2d = bm.createGraphics()!!
    }

    fun add(w: Int, h: Int, color: Color): Boolean {
        if (fails >= 30) return false
        val rect = Rect2i(0, 0, w, h)
        val start = GLFW.glfwGetTime()
        val result = packer.addRect(rect)
        time += GLFW.glfwGetTime() - start
        if (result) {
            g2d.color = color
            g2d.fillRect(rect.x(), rect.y(), w, h)
            ++n
        } else {
            ++fails
        }
        return true
    }

    fun print() {
        println("$n rectangles")
        println("Coverage: " + packer.coverage + " (higher is better)")
        println("" + (time * 1.0E9 / n).toInt() + " nanoseconds per rectangle (lower is better)")
    }
}
